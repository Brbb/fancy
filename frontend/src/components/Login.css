.Login {
  color: cornflowerblue;
  vertical-align: middle;
  text-align: center;
  height: 350px;
  width: 350px;
  background: white;
  position: absolute;
  top: 30%;
  left: 35%;
}

.form {
  display: flex;
  flex-direction: column;
  height: 100%;
  align-items: center;
}
