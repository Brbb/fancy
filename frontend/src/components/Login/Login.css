.Login {
  height: 350px;
  width: 350px;
}


.centered {
  position: fixed;
  top: 40%;
  left: 50%;
  transform: translate(-50%, -40%);
}