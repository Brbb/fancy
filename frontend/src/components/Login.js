import React, { Component } from "react";
import { Link } from "react-router-dom";
import "./Login.css";
import auth from "../services/auth/api";

class Login extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      error: ""
    };
  }

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };

  login = async () => {
    var loginResult = await auth.login(
      this.state.username,
      this.state.password
    ); // SSL
    if (!loginResult.err) {
      this.props.history.push("/settings");
    } else {
      this.setState({ error: loginResult.err });
    }
  };

  render() {
    return (
      <div className="Login formContainer">
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="formTitle">Fancy App</label>
          <label className="textFieldLabel">Email</label>
          <input
            type="text"
            className="textField"
            placeholder="user@email.com"
            data-test="username"
            value={this.state.username}
            onChange={this.handleUserChange}
          />

          <label className="textFieldLabel">Password</label>
          <input
            type="password"
            placeholder="super secret"
            data-test="password"
            className="textField"
            value={this.state.password}
            onChange={this.handlePassChange}
          />
          {this.state.error && (
            <label
              className="loginError"
              data-test="error"
              onClick={this.dismissError}
            >
              {this.state.error}
            </label>
          )}
          <input
            className="submitBtn"
            type="button"
            value="Log In"
            onClick={this.login}
            disabled={this.state.username === "" || this.state.password === ""}
          />
          <button className="genericBtn">
            <Link to="/new">Create New Account</Link>
          </button>
        </form>
      </div>
    );
  }
}

export default Login;
