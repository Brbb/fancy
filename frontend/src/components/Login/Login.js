import React, { Component } from "react";
import "./Login.css";
import authApi from "../../services/auth/api";
import userApi from "../../services/users/api";
import {
  Button,
  InputField,
  Message,
  SectionTitle
} from "../Elements/Basics";

class Login extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      password: "",
      error: ""
    };

    //this.loadUser = this.loadUser.bind(this);
    this.login = this.login.bind(this);
  }

  loadUser = async userId => {
    let user = await userApi.getById(userId);
    if (!user.err) {
      this.props.history.push({
        pathname: "/settings",
        state: { user: user }
      });
    } else {
      this.props.history.push("/error");
    }
  };

  dismissError = () => {
    this.setState({ error: "" });
  };

  handleUserChange = event => {
    this.setState({ username: event.target.value });
  };
  handlePassChange = event => {
    this.setState({ password: event.target.value });
  };

  async login() {
    var loginResult = await authApi.login(
      this.state.username,
      this.state.password
    ); // SSL
    if (!loginResult.err) {
      localStorage.setItem("jwt", loginResult.token);
      localStorage.setItem("me", loginResult.userid);
      await this.loadUser(loginResult.userId);
    } else {
      this.setState({ error: loginResult.err });
    }
  }

  render() {
    return (
      <div className="f-section bordered centered">
        <SectionTitle text="Login" />
        <InputField
          placeholder="user@email.com"
          value={this.state.username}
          onChange={this.handleUserChange}
          text="Email"
          name="emailField"
        />
        <InputField
          type="password"
          placeholder="super secret"
          value={this.state.password}
          onChange={this.handlePassChange}
          text="Password"
          name="passwordField"
        />

        <Button
          name="login-button"
          className="f-submit-button"
          text="Login"
          onClick={this.login}
          disabled={this.state.username === "" || this.state.password === ""}
        />

        <Button
          name="create-button"
          className="f-button"
          text="Create new account"
          onClick={() => {
            this.props.history.push("/new");
          }}
        />

        <Message text={this.state.error} success={this.state.error === ""} />
      </div>
    );
  }
}

export default Login;
