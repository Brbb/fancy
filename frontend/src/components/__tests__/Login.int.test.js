import Login from "../Login/Login";
import axios from "axios";
import userApi from "../../services/users/api";
import authApi from "../../services/auth/api";
jest.mock("axios");

const mockUser = email => {
  return {
    data: {
      id: "asd123",
      email: email,
      settings: {
        language: "English",
        isPrivate: false
      }
    }
  };
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe("Login component, integration test", () => {
  describe("Login successful", () => {
    test("User types email and passwords, clicks on login and goes to /settings.", () => {
      let location = {};
      let historyMock = {
        push: jest.fn(payload => {
          location = payload;
        })
      };

      jest.spyOn(axios, "post").mockImplementation(() => {
        return Promise.resolve({
          data: {
            userId: mockUser().data.id,
            token: "any-valid-token"
          }
        });
      });

      jest.spyOn(axios, "get").mockImplementation(url => {
        return Promise.resolve(mockUser());
      });

      jest.spyOn(userApi, "getById");
      jest.spyOn(authApi, "login");

      let wrapper = mount(<Login history={historyMock} />);

      wrapper
        .find('[name="emailField"]')
        .last()
        .simulate("change", { target: { value: "test@email.com" } });
      wrapper
        .find('[name="passwordField"]')
        .last()
        .simulate("change", { target: { value: "pass123" } });

      let loginButton = wrapper.find('[name="login-button"]').first();

      expect(wrapper).toMatchSnapshot(); // fields are populated and button is ready to be clicked
      expect(wrapper.state().username).toEqual("test@email.com");
      expect(wrapper.state().password).toEqual("pass123");
      expect(loginButton.prop("disabled")).toBeFalsy();
      expect(axios.post).toBeCalledTimes(0);
      expect(axios.get).toBeCalledTimes(0);

      // submit
      return loginButton
        .prop("onClick")()
        .then(() => {
          // we called the login API
          expect(axios.post).toBeCalledTimes(1);
          expect(authApi.login).toBeCalledWith(
            wrapper.state().username,
            wrapper.state().password
          );

          // we called then the user API to retrieve the user by id with the previous result
          expect(userApi.getById).toBeCalledTimes(1);
          expect(userApi.getById).toBeCalledWith(mockUser().data.id);
          expect(axios.get).toBeCalledTimes(1);

          // successfully redirected to setting with user as prop
          expect(location.pathname).toBe("/settings");
          expect(location.state.user.id).toBe(mockUser().data.id);
        });
    });
  });
});
