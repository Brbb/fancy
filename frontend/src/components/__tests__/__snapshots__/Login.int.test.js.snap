// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component, integration test Login successful User types email and passwords, clicks on login and goes to /settings. 1`] = `
<div
  className="f-section bordered centered"
>
  <div
    className="f-section-title"
  >
    <label>
      Login
    </label>
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Email
    </label>
    <input
      className="f-input-field"
      name="emailField"
      onChange={[Function]}
      placeholder="user@email.com"
      value="test@email.com"
    />
  </div>
  <div
    className="f-input"
  >
    <label
      className="f-input-field-label"
    >
      Password
    </label>
    <input
      className="f-input-field"
      name="passwordField"
      onChange={[Function]}
      placeholder="super secret"
      type="password"
      value="pass123"
    />
  </div>
  <input
    className="f-submit-button"
    disabled={false}
    name="login-button"
    onClick={[Function]}
    type="button"
    value="Login"
  />
  <input
    className="f-button"
    name="create-button"
    onClick={[Function]}
    type="button"
    value="Create new account"
  />
  <label
    className="f-message f-message-s"
  />
</div>
`;
